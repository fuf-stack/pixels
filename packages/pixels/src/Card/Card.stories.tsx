import type { Meta, StoryObj } from '@storybook/react-vite';

import Card from './Card';

const meta: Meta<typeof Card> = {
  title: 'pixels/Card',
  component: Card,
};

export default meta;
type Story = StoryObj<typeof Card>;

export const Default: Story = {
  args: {
    children: (
      <p className="w-96">
        Lorem commodo nulla adipisicing cillum Lorem sunt laboris exercitation
        esse. Exercitation ut cillum cupidatat deserunt occaecat pariatur
        laborum ut reprehenderit veniam. Culpa fugiat aliqua consectetur sit
        esse.
      </p>
    ),
  },
};

export const FooterAndHeader: Story = {
  args: {
    footer: 'Footer',
    header: 'Header',
    children: (
      <p className="w-96">
        Lorem commodo nulla adipisicing cillum Lorem sunt laboris exercitation
        esse. Exercitation ut cillum cupidatat deserunt occaecat pariatur
        laborum ut reprehenderit veniam. Culpa fugiat aliqua consectetur sit
        esse.
      </p>
    ),
  },
};

export const CustomSlotStyles: Story = {
  args: {
    footer: 'Footer',
    header: 'Header',
    children: (
      <p className="w-96">
        Lorem commodo nulla adipisicing cillum Lorem sunt laboris exercitatio
      </p>
    ),
    className: {
      base: 'text-blue-400',
      body: 'text-green-400',
      divider: 'bg-yellow-400',
    },
  },
};
