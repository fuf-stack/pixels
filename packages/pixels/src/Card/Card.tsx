import type { TVClassName, TVProps } from '@fuf-stack/pixel-utils';
import type { ReactNode } from 'react';

import {
  Card as NextCard,
  CardBody as NextCardBody,
  CardFooter as NextCardFooter,
  CardHeader as NextCardHeader,
} from '@nextui-org/card';
import { Divider as NextDivider } from '@nextui-org/divider';

import { tv, variantsToClassNames } from '@fuf-stack/pixel-utils';

// card styling variants
export const cardVariants = tv({
  slots: {
    base: 'border border-divider',
    body: '',
    divider: 'my-0 border-divider',
    footer: '',
    header: 'text-base font-semibold',
  },
});

type VariantProps = TVProps<typeof cardVariants>;
type ClassName = TVClassName<typeof cardVariants>;

export interface CardProps extends VariantProps {
  /** card body content */
  children?: ReactNode;
  /** CSS class name */
  className?: ClassName;
  /** footer content */
  footer?: ReactNode;
  /** header content */
  header?: ReactNode;
  /** ref */
  ref?: React.RefObject<HTMLDivElement | null> | undefined;
  /** HTML data-testid attribute used in e2e tests */
  testId?: string;
}

/**
 * Card component based on [NextUI Card](https://nextui.org/docs/components/card)
 */
const Card = ({
  children = null,
  className = undefined,
  footer = undefined,
  header = undefined,
  ref = undefined,
  testId = undefined,

}: CardProps) => {
  // classNames from slots
  const variants = cardVariants();
  const { divider: dividerClassName, ...classNames } = variantsToClassNames(
    variants,
    className,
    'base',
  );

  const divider = <NextDivider className={dividerClassName} />;

  return (
    <NextCard
      classNames={classNames}
      data-testid={testId && `card_${testId}`}
      fullWidth
      radius="sm"
      ref={ref}
      shadow="none"
    >
      {header && (
        <>
          <NextCardHeader data-testid={testId && `card_header_${testId}`}>
            {header}
          </NextCardHeader>
          {divider}
        </>
      )}
      <NextCardBody data-testid={testId && `card_body_${testId}`}>
        {children}
      </NextCardBody>
      {footer && (
        <>
          {divider}
          <NextCardFooter data-testid={testId && `card_footer_${testId}`}>
            {footer}
          </NextCardFooter>
        </>
      )}
    </NextCard>
  );
};

export default Card;
