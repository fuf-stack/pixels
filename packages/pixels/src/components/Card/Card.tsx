import type { ReactNode } from 'react';
import type { VariantProps } from 'tailwind-variants';

import {
  Card as NextCard,
  CardBody as NextCardBody,
  CardFooter as NextCardFooter,
  CardHeader as NextCardHeader,
} from '@nextui-org/card';
import { Divider as NextDivider } from '@nextui-org/divider';
import createDebug from 'debug';
import { tv } from 'tailwind-variants';

const debug = createDebug('component:Card');

// card styling variants
export const cardVariants = tv({
  slots: {
    base: 'border border-slate-300',
    body: '',
    divider: 'my-0 bg-slate-300',
    footer: '',
    header: 'text-base font-semibold',
  },
});

type CardVariantProps = VariantProps<typeof cardVariants>;
type CardVariantSlots = Partial<
  Record<keyof ReturnType<typeof cardVariants>, string>
>;

export interface CardProps extends CardVariantProps {
  /** child components */
  children?: ReactNode;
  /** CSS class name */
  className?: string | CardVariantSlots;
  /** footer content */
  footer?: ReactNode;
  /** header content */
  header?: ReactNode;
  /** HTML data-testid attribute used in e2e tests */
  testId?: string;
}

/**
 * Card component based on [NextUI Card](https://nextui.org/docs/components/card)
 */
const Card = ({
  children = null,
  className = undefined,
  testId = undefined,
  header = undefined,
  footer = undefined,
}: CardProps) => {
  debug('Card', { className, testId });
  const {
    base: baseSlot,
    body: bodySlot,
    divider: dividerSlot,
    footer: footerSlot,
    header: headerSlot,
  } = cardVariants();

  return (
    <NextCard
      data-testid={testId && `card_${testId}`}
      className={baseSlot({
        className: typeof className === 'object' ? className.base : className,
      })}
      fullWidth
      radius="sm"
      shadow="none"
    >
      {header && (
        <>
          <NextCardHeader
            data-testid={testId && `card_header_${testId}`}
            className={headerSlot({
              className: typeof className === 'object' && className.header,
            })}
          >
            {header}
          </NextCardHeader>
          <NextDivider
            className={dividerSlot({
              className: typeof className === 'object' && className.divider,
            })}
          />
        </>
      )}
      <NextCardBody
        data-testid={testId && `card_body_${testId}`}
        className={bodySlot({
          className: typeof className === 'object' && className.body,
        })}
      >
        {children}
      </NextCardBody>
      {footer && (
        <>
          <NextDivider
            className={dividerSlot({
              className: typeof className === 'object' && className.divider,
            })}
          />
          <NextCardFooter
            data-testid={testId && `card_footer_${testId}`}
            className={footerSlot({
              className: typeof className === 'object' && className.footer,
            })}
          >
            {footer}
          </NextCardFooter>
        </>
      )}
    </NextCard>
  );
};

export default Card;
