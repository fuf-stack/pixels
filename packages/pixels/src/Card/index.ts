import Card, { cardVariants } from './Card';

export type { CardProps } from './Card';

export { Card, cardVariants };

export default Card;
