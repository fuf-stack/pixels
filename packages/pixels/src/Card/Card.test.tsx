import { describe, expect, test } from 'vitest';

import storySnapshots from '@repo/storybook-config/story-snapshots';
import { render } from '@testing-library/react';

import Card from './Card';
import * as stories from './Card.stories';

describe('Story Snapshots', () => {
  storySnapshots(stories);
});

describe('Coverage', () => {
  test('should render with no children', () => {
    const { container } = render(<Card />);
    expect(container.firstChild).toMatchSnapshot();
  });
  test('Should have correct testIds', () => {
    const { container } = render(<Card testId="card1" />);
    expect(container.firstChild).toMatchSnapshot();
  });
  test('should render with only header', () => {
    const { container } = render(
      <Card footer="id" header={<div>test</div>} testId="testId" />,
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});
