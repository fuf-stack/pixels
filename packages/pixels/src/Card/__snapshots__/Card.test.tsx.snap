// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Coverage > Should have correct testIds 1`] = `
<div
  class="flex flex-col relative overflow-hidden h-auto text-foreground box-border bg-content1 outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 shadow-none rounded-small w-full transition-transform-background motion-reduce:transition-none border border-divider"
  data-testid="card_card1"
  tabindex="-1"
>
  <div
    class="relative flex w-full p-3 flex-auto flex-col place-content-inherit align-items-inherit h-auto break-words text-left overflow-y-auto subpixel-antialiased"
    data-testid="card_body_card1"
  />
</div>
`;

exports[`Coverage > should render with no children 1`] = `
<div
  class="flex flex-col relative overflow-hidden h-auto text-foreground box-border bg-content1 outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 shadow-none rounded-small w-full transition-transform-background motion-reduce:transition-none border border-divider"
  tabindex="-1"
>
  <div
    class="relative flex w-full p-3 flex-auto flex-col place-content-inherit align-items-inherit h-auto break-words text-left overflow-y-auto subpixel-antialiased"
  />
</div>
`;

exports[`Coverage > should render with only header 1`] = `
<div
  class="flex flex-col relative overflow-hidden h-auto text-foreground box-border bg-content1 outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 shadow-none rounded-small w-full transition-transform-background motion-reduce:transition-none border border-divider"
  data-testid="card_testId"
  tabindex="-1"
>
  <div
    class="flex p-3 z-10 w-full justify-start items-center shrink-0 overflow-inherit color-inherit subpixel-antialiased rounded-t-small text-base font-semibold"
    data-testid="card_header_testId"
  >
    <div>
      test
    </div>
  </div>
  <hr
    class="shrink-0 bg-divider border-none w-full h-divider my-0 border-divider"
    role="separator"
  />
  <div
    class="relative flex w-full p-3 flex-auto flex-col place-content-inherit align-items-inherit h-auto break-words text-left overflow-y-auto subpixel-antialiased"
    data-testid="card_body_testId"
  />
  <hr
    class="shrink-0 bg-divider border-none w-full h-divider my-0 border-divider"
    role="separator"
  />
  <div
    class="p-3 h-auto flex w-full items-center overflow-hidden color-inherit subpixel-antialiased rounded-b-small"
    data-testid="card_footer_testId"
  >
    id
  </div>
</div>
`;

exports[`Story Snapshots > CustomSlotStyles 1`] = `
<div>
  <div
    class="flex flex-col relative overflow-hidden h-auto box-border bg-content1 outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 shadow-none rounded-small w-full transition-transform-background motion-reduce:transition-none border border-divider text-blue-400"
    tabindex="-1"
  >
    <div
      class="flex p-3 z-10 w-full justify-start items-center shrink-0 overflow-inherit color-inherit subpixel-antialiased rounded-t-small text-base font-semibold"
    >
      Header
    </div>
    <hr
      class="shrink-0 border-none w-full h-divider my-0 border-divider bg-yellow-400"
      role="separator"
    />
    <div
      class="relative flex w-full p-3 flex-auto flex-col place-content-inherit align-items-inherit h-auto break-words text-left overflow-y-auto subpixel-antialiased text-green-400"
    >
      <p
        class="w-96"
      >
        Lorem commodo nulla adipisicing cillum Lorem sunt laboris exercitatio
      </p>
    </div>
    <hr
      class="shrink-0 border-none w-full h-divider my-0 border-divider bg-yellow-400"
      role="separator"
    />
    <div
      class="p-3 h-auto flex w-full items-center overflow-hidden color-inherit subpixel-antialiased rounded-b-small"
    >
      Footer
    </div>
  </div>
</div>
`;

exports[`Story Snapshots > Default 1`] = `
<div>
  <div
    class="flex flex-col relative overflow-hidden h-auto text-foreground box-border bg-content1 outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 shadow-none rounded-small w-full transition-transform-background motion-reduce:transition-none border border-divider"
    tabindex="-1"
  >
    <div
      class="relative flex w-full p-3 flex-auto flex-col place-content-inherit align-items-inherit h-auto break-words text-left overflow-y-auto subpixel-antialiased"
    >
      <p
        class="w-96"
      >
        Lorem commodo nulla adipisicing cillum Lorem sunt laboris exercitation esse. Exercitation ut cillum cupidatat deserunt occaecat pariatur laborum ut reprehenderit veniam. Culpa fugiat aliqua consectetur sit esse.
      </p>
    </div>
  </div>
</div>
`;

exports[`Story Snapshots > FooterAndHeader 1`] = `
<div>
  <div
    class="flex flex-col relative overflow-hidden h-auto text-foreground box-border bg-content1 outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 shadow-none rounded-small w-full transition-transform-background motion-reduce:transition-none border border-divider"
    tabindex="-1"
  >
    <div
      class="flex p-3 z-10 w-full justify-start items-center shrink-0 overflow-inherit color-inherit subpixel-antialiased rounded-t-small text-base font-semibold"
    >
      Header
    </div>
    <hr
      class="shrink-0 bg-divider border-none w-full h-divider my-0 border-divider"
      role="separator"
    />
    <div
      class="relative flex w-full p-3 flex-auto flex-col place-content-inherit align-items-inherit h-auto break-words text-left overflow-y-auto subpixel-antialiased"
    >
      <p
        class="w-96"
      >
        Lorem commodo nulla adipisicing cillum Lorem sunt laboris exercitation esse. Exercitation ut cillum cupidatat deserunt occaecat pariatur laborum ut reprehenderit veniam. Culpa fugiat aliqua consectetur sit esse.
      </p>
    </div>
    <hr
      class="shrink-0 bg-divider border-none w-full h-divider my-0 border-divider"
      role="separator"
    />
    <div
      class="p-3 h-auto flex w-full items-center overflow-hidden color-inherit subpixel-antialiased rounded-b-small"
    >
      Footer
    </div>
  </div>
</div>
`;
